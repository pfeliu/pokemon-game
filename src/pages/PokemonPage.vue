<template>
  <h1 v-if="!pokemon">Wait, please</h1>
  <div v-else>
    <h1>Who is this pok√©mon?</h1>
    <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon" />

    <PokemonOptions :pokemons="pokemonArr" @selection="checkAnswer" />

    <template v-if="showAnswer" >
      <h2 class="fade-in">{{ message }}</h2>

      <button @click="newGame">New game</button>
    </template>
  </div>
</template>

<script>
import PokemonPicture from '@/components/PokemonPicture';
import PokemonOptions from '@/components/PokemonOptions';
import getPokemonOptions from '@/helpers/getPokemonOptions';

export default {
  components: { PokemonPicture, PokemonOptions },
  data() {
    return {
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: '',
    };
  },
  methods: {
    async mixPokemonArray() {
      this.pokemonArr = await getPokemonOptions();

      const rndInt = Math.floor(Math.random() * 4);
      this.pokemon = this.pokemonArr[rndInt];
    },
    checkAnswer(selectedId) {
      this.showPokemon = true;
      this.showAnswer = true;

      if (selectedId === this.pokemon.id) {
        this.message = `Well done!!, it's ${this.pokemon.name}`;
      } else {
        this.message = `Ooops!!! no!!!, was ${this.pokemon.name}`;
      }
    },
    newGame() {
        this.pokemon = null
        this.showPokemon = false
        this.showAnswer = false
        this.pokemonArr = []
        this.mixPokemonArray()
    }
  },
  mounted() {
    this.mixPokemonArray();
  },
};
</script>
