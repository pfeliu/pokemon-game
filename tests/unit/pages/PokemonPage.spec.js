import PokemonPage from '@/pages/PokemonPage.vue';
import { mount, shallowMount } from '@vue/test-utils';
import { pokemons } from '../mocks/pokemons.mocks';
import { wrap } from 'module';

describe('Pokemon page', () => {
  let wrapper;
  //let mixPokemonArraySpy;

  beforeEach(() => {
    // mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray');
    wrapper = shallowMount(PokemonPage);
  });
  test('should match eith snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should call mixedPoekmon array on mounted', () => {
    const mixPokemonArraySpy = jest.spyOn(
      PokemonPage.methods,
      'mixPokemonArray'
    );
    const wrapper = shallowMount(PokemonPage);
    expect(mixPokemonArraySpy).toHaveBeenCalled();
  });

  test('should match with snapshot when load data', () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: '',
        };
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should show components Pokemon picture and options', () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: '',
        };
      },
    });

    const picture = wrapper.find('pokemon-picture-stub');
    const options = wrapper.find('pokemon-options-stub');

    expect(picture.exists()).toBeTruthy();
    expect(options.exists()).toBeTruthy();

    expect(picture.attributes('pokemonid')).toBe('11');

    expect(options.attributes('pokemons')).toBeTruthy();
  });

  test('should check answer method', async () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: '',
        };
      },
    });

    await wrapper.vm.checkAnswer(11);

    expect(wrapper.find('h2').exists()).toBeTruthy();

    expect(wrapper.vm.showPokemon).toBe(true);

    expect(wrapper.find('h2').text()).toBe(
      `Well done!!, it's ${pokemons[0].name}`
    );

    await wrapper.vm.checkAnswer(1);

    expect(wrapper.vm.message).toBe(`Ooops!!! no!!!, was ${pokemons[0].name}`);
  });
});
