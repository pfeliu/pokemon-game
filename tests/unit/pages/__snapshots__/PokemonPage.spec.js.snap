// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemon page should match eith snapshot 1`] = `<h1>Wait, please</h1>`;

exports[`Pokemon page should match with snapshot when load data 1`] = `
<div>
  <h1>Who is this pok√©mon?</h1>
  <pokemon-picture-stub pokemonid="11" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
